docs/Middleware-Guide.md

```diff
@@ -0,0 +1,82 @@
+1. The staker first has to queue its withdrawal request with EigenLayer. The staker can place this request by calling  `queueWithdrawal(..)` in the EigenLayer's `StrategyManager.sol`. On successful call to this, the event `WithdrawalQueued(..)` is emitted.
+2. The staker then has to notify all the middlewares that its operator has opted in about its intention to partially withdraw some of its stake or completely stop participating and also record the latest time in the EigenLayer contracts until which the staker is obliged to participate with the current stake. To do this, the staker calls on the `updateStakes(..)` function in the Registry contract that calls `recordStakeUpdate(..)` in the middleware's ServiceManager contract which in turn accesses `recordStakeUpdate(..)` in the EigenLayer's Slasher contract.  On successful execution of this call, the event `MiddlewareTimesAdded(..)` is emitted.
+3. After the latest time until which the staker is obliged to participate with the current stake passes, the staker now can complete its queued withdrawal by calling on the `completeQueuedWithdrawal(..)` function in `StrategyManager.sol`.
+
+The following figure illustrates the above flow: 
+![Staker deregistering](./images/staker_withdrawing.png)
+
+
+### *Slashing*
+As mentioned above, EigenLayer is built to support slashing as a result of an on-chain-checkable, objectively attributable action. In order for a middleware to be able to slash an operator in an objective manner, the middleware needs to deploy a DisputeResolution contract where anyone would be able to raise a challenge against an EigenLayer operator for its adversarial action. On successful challenge, the DisputeResolution contract calls the `freezeOperator(..)` function in ServiceManager contract which in turn accesses the `freezeOperator(..)` function in the EigenLayer's Slasher contract. On successful execution of this API, the event `OperatorFrozen(..)` is emitted.  
+
+The following figure illustrates the above flow: 
+![Slashing](./images/slashing.png)
+
+
+
+
+## Guide To Provided Middleware Contracts:
+The EigenLayer team has built a set of reusable and extensible contracts for use in middlewares built on top of EigenLayer. These are contained in the general-purpose /middleware/ folder, which contains code that can be extended, used directly, or consulted as a reference in building middleware on top of EigenLayer. There are several basic contracts that all middleware-specific contracts can be built on:
+- The *RegistryBase contract* is a basic registry contract that can be used to track operators opted into running a middleware.  Importantly, this base registry contract assumes a maximum of two quorums, where each quorum represents an aggregation of a certain type of stake.  Middlewares may want to weigh their rewards based on the type of stake delegated to their operators and weigh each quorum differently.  
```

Comment:
since the RegistryBase inherits *from* the VoteWeigherBase, we might want to swap the order of these two contracts, so we go from least to most derived (rather than hopping back and forth in the inheritance tree)

Reply:
done: